LSM: use inline hash for Keys

kprotty opened the pull request:
Wyhash on opaque bytes takes up a visible amount of the profile. We can utilize the statically-known size of our hash inputs to inline it and remove all branching.

Using `@call(.{ .modifier = .always_inline }, Wyhash, args)` didn't seem to be enough to force all its branching to be inlined everywhere. So instead, I opted to implement Google Abseil's [low_level_hash](https://github.com/abseil/abseil-cpp/blob/master/absl/hash/internal/low_level_hash.cc) which is a simplified version of Wyhash and is [endorsed by wyhash](https://github.com/wangyi-fudan/wyhash/blob/77e50f267fbc7b8e2d09f2d455219adb70ad4749/README.md?plain=1#L83).

## Pre-merge checklist

Performance:

* [x] Compare `zig build benchmark` on linux before and after this pr.
    ``` sh
    # benchmark results before
    1224 batches in 92.44 s
    load offered = 1000000 tx/s
    load accepted = 108178 tx/s
    batch latency p00 = 0 ms
    batch latency p10 = 31 ms
    batch latency p20 = 32 ms
    batch latency p30 = 33 ms
    batch latency p40 = 34 ms
    batch latency p50 = 35 ms
    batch latency p60 = 36 ms
    batch latency p70 = 36 ms
    batch latency p80 = 37 ms
    batch latency p90 = 38 ms
    batch latency p100 = 4170 ms
    transfer latency p00 = 0 ms
    transfer latency p10 = 4306 ms
    transfer latency p20 = 9067 ms
    transfer latency p30 = 16122 ms
    transfer latency p40 = 24505 ms
    transfer latency p50 = 34398 ms
    transfer latency p60 = 45255 ms
    transfer latency p70 = 55560 ms
    transfer latency p80 = 65938 ms
    transfer latency p90 = 75036 ms
    transfer latency p100 = 82452 ms
    
    # benchmark results after
    1225 batches in 82.36 s
    load offered = 1000000 tx/s
    load accepted = 121417 tx/s
    batch latency p00 = 0 ms
    batch latency p10 = 23 ms
    batch latency p20 = 24 ms
    batch latency p30 = 24 ms
    batch latency p40 = 25 ms
    batch latency p50 = 27 ms
    batch latency p60 = 28 ms
    batch latency p70 = 29 ms
    batch latency p80 = 29 ms
    batch latency p90 = 30 ms
    batch latency p100 = 4253 ms
    transfer latency p00 = 0 ms
    transfer latency p10 = 3493 ms
    transfer latency p20 = 7083 ms
    transfer latency p30 = 13025 ms
    transfer latency p40 = 20471 ms
    transfer latency p50 = 29355 ms
    transfer latency p60 = 38708 ms
    transfer latency p70 = 48141 ms
    transfer latency p80 = 57847 ms
    transfer latency p90 = 66147 ms
    transfer latency p100 = 72372 ms
    ```

kprotty committed:
lsm: use hash_inline for statically known Keys

This turns into an O(1) operation instead of calling
into Wyhash to hash a bunch of bytes.

kprotty committed:
stdx: simplify hash_inline

cb22 commented on src/stdx.zig:
```diff
@@ -202,3 +202,33 @@ pub const log = if (builtin.is_test)
+
+/// Inline version of Google Abseil "LowLevelHash" (inspired by wyhash).
+/// https://github.com/abseil/abseil-cpp/blob/master/absl/hash/internal/low_level_hash.cc
+pub inline fn hash_inline(value: anytype) u64 {
```
I think there might be a bit of a footgun opportunity here with `value: anytype` - eg, `hash_inline("hello world")` won't do what you expect.

cb22 replied:
Other than that - perhaps some hardcoded tests to compare to Abseil output?

kprotty committed:
stdx: add test vectors for low_level_hash

Also guard against trying to statically hash
invalid types like slices and pointers.
